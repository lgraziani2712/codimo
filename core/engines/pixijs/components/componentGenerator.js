/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import { type Container } from 'pixi.js';

type Functionality = Object;
export type CodimoComponent = {
  view: Container,

  [key: string]: any,
};
export type FunctionalityBuilder = (
  size: number,
  margin: number,
  component: CodimoComponent,
) => Functionality;
export type Generator = {|
  addFunctionality(key: string, functionality: FunctionalityBuilder): Generator,
  build(): CodimoComponent,
|};

/**
 * This generator creates components with functionalities.
 *
 * @todo Add example
 * @version 1.0.0
 * @param {Container} view A PixiJS Container.
 * @param {number} size Block's size.
 * @param {number} margin CÃ³dimo's block margin.
 * @return {Generator} The component generator object for  this specific `view`.
 */
export default function componentGenerator(
  view: Container,
  size: number,
  margin: number,
): Generator {
  const functionalities = new Map();

  return {
    /**
     * Sets a `functionality` into a Map collection on `key`.
     * This makes replacement easy: pass a new function with the same
     * `key` as the one you want to replace.
     *
     * @param {string} key Unique key.
     * @param {FunctionalityBuilder} functionality A new animation or action for
     *                                             the component.
     * @return {Generator} This allow chaining calls.
     */
    addFunctionality(key: string, functionality: FunctionalityBuilder): Generator {
      functionalities.set(key, functionality);

      return this;
    },
    /**
     * It instanciate every functionality and adds it to the component
     * object.
     *
     * Also sets some default props to the view.
     *
     * @return {CodimoComponent} A component with a set of functionalities.
     */
    build() {
      const functionalityValues = functionalities.values();
      let component = {
        view,
      };

      for (const functionality of functionalityValues) {
        component = {
          ...component,
          ...functionality(size, margin, component),
        };
      }

      return component;
    },
  };
}
