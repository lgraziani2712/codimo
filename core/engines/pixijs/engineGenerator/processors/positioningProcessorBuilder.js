/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import { type Instruction } from 'core/workspaces/blockly/parseInstructions';
import {
  MOVE_FORWARD,
  MOVE_RIGHT,
  MOVE_BACKWARD,
  MOVE_LEFT,
} from 'core/constants/instructions';
import {
  type CodimoComponent,
} from 'core/engines/pixijs/components/componentGenerator';

import { type ExecutionProcessor } from './processorGenerator';

export type PositioningState = {|
  instruction: string,
  oldPosition: string,
  newPosition: string,
|};

type BeforeUpdateStateChecker = (state: PositioningState) => Promise<void | Error>;
type CheckersCollection = Map<string, BeforeUpdateStateChecker>;

/* eslint-disable no-magic-numbers */
const directions = {
  [MOVE_FORWARD]: [0, -1],
  [MOVE_RIGHT]: [1, 0],
  [MOVE_BACKWARD]: [0, 1],
  [MOVE_LEFT]: [-1, 0],
};
/* eslint-enable */

/**
 * This ExecutionProcessor updates the component's position
 * while animating it.
 *
 * @todo Add example
 * @version 1.0.0
 * @param  {CodimoComponent}    component The component that will have attached the processor.
 * @param  {CheckersCollection} checkers  A collection of Checker objects.
 * @return {ExecutionProcessor}           The new instance.
 */
const positioningExecutionProcessorBuilder = (
  component: CodimoComponent,
  checkers: CheckersCollection,
): ExecutionProcessor => {
  if (
    typeof component.position !== 'string' ||
    typeof component.updatePosition !== 'function'
  ) {
    throw new Error(
      // eslint-disable-next-line max-len
      'The `positioning` engine\'s processor requires the component to have the `positioning` functionality',
    );
  }

  return {
    async instructionProcessor(instruction: Instruction) {
      const direction = instruction.key;
      const times = parseInt(instruction.params[0]);

      if (!directions.hasOwnProperty(direction)) {
        return;
      }
      for (let i = 0; i < times; i++) {
        const oldPosition = component.position;
        const newPosition =
          oldPosition
              .split(',')
              .map((pos, idx) => (parseInt(pos) + directions[direction][idx]))
              .join(',');

        for (const checker of checkers.values()) {
          await checker({
            instruction: direction,
            oldPosition,
            newPosition,
          });
        }

        await component.updatePosition(newPosition);
      }
    },
  };
};

export default positioningExecutionProcessorBuilder;
