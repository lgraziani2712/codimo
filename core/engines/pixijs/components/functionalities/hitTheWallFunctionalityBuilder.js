/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import { TimelineLite, SlowMo } from 'gsap';

import {
  ZERO,
  HALF,
  DEFAULT_MOVEMENT_DURATION,
} from 'core/constants/numbers';

import { type FunctionalityBuilder } from '../componentGenerator';
import { type Direction } from '../blockGenerator';

const HIT_DURATION = 0.4;
const EASE_STRENG_AT_EACH_END = ZERO;

/**
 * A builder with two prerequired params.
 *
 * @version 1.0.0
 * @param  {number} [movementDuration=0.5] Main animation time
 * @param  {number} [hitDuration=0.4]      Hit effect time
 * @return {FunctionalityBuilder}          This new function returns an object with
 *                                         the `hitTheWall` functionality.
 */
const hitTheWallFunctionalityBuilder = (
  movementDuration: number = DEFAULT_MOVEMENT_DURATION,
  hitDuration: number = HIT_DURATION,
): FunctionalityBuilder => (
  size: number,
  margin: number,
) => ({
  /**
   * This animation simulates as if the component has hit a wall.
   * It uses a Linear ease.
   *
   * @param  {Direction} direction what wall has hit.
   * @return {Promise<void>}       animation's promise.
   */
  hitTheWall(direction: Direction) {
    return new Promise((onComplete) => {
      let movement;
      const distance = size / HALF - margin;
      const oldValues = {
        x: this.view.x,
        y: this.view.y,
      };
      const timeline = new TimelineLite({
        // FIXME monkeypatch @see https://github.com/greensock/GreenSock-JS/issues/226
        onComplete: () => {
          this.view.x = oldValues.x;
          this.view.y = oldValues.y;

          onComplete();
        },
      });

      switch (direction) {
      case 'top':
        movement = {
          x: this.view.x,
          y: this.view.y - distance,
        };
        break;
      case 'right':
        movement = {
          x: this.view.x + distance,
          y: this.view.y,
        };
        break;
      case 'bottom':
        movement = {
          x: this.view.x,
          y: this.view.y + distance,
        };
        break;
      case 'left':
        movement = {
          x: this.view.x - distance,
          y: this.view.y,
        };
        break;
      }

      timeline
          .to(this.view, movementDuration, {
            ...movement,
            ease: SlowMo.ease.config(hitDuration, EASE_STRENG_AT_EACH_END, true),
          })
          .to(this.view, movementDuration, {
            ...movement,
            ease: SlowMo.ease.config(hitDuration, EASE_STRENG_AT_EACH_END, true),
          })
          .to(this.view, movementDuration, {
            ...movement,
            ease: SlowMo.ease.config(hitDuration, EASE_STRENG_AT_EACH_END, true),
          });
    });
  },
});

export default hitTheWallFunctionalityBuilder;
