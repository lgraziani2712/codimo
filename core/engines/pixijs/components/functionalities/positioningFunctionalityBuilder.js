/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import { TweenLite, Linear } from 'gsap';

import { HALF, DEFAULT_MOVEMENT_DURATION } from 'core/constants/numbers';

import { type CodimoComponent, type FunctionalityBuilder } from '../componentGenerator';

/**
 * A builder with two prerequired params.
 *
 * @version 1.0.0
 * @param  {string} position               CodimoComponent's starting point.
 * @param  {string} [endPosition]          CodimoComponent's possible end point.
 * @param  {number} [movementDuration=0.5] Main animation time
 * @return {FunctionalityBuilder}          This new function returns an object with
 *                                         the `positioning` functionality.
 */
const positioningFunctionalityBuilder = (
  position: string,
  endPosition: string = '',
  movementDuration: number = DEFAULT_MOVEMENT_DURATION,
): FunctionalityBuilder => (
  size: number,
  margin: number,
  component: CodimoComponent,
) => {
  if (!component.view.parent) {
    throw new Error('`positioning` functionality requires the component to have a parent');
  }
  const initialParent = component.view.parent;
  const initialPosition =
    position
      .split(',')
      .map((string: string): number => (parseInt(string)));

  component.view.x = initialPosition[0] * (size + margin) + size / HALF + margin;
  component.view.y = initialPosition[1] * (size + margin) + size / HALF + margin;

  return {
    /**
     * Actual position
     *
     * @readonly
     */
    position,
    /**
     *
     * @readonly
     */
    initialPosition,
    /**
     *
     * @readonly
     */
    endPosition,
    /**
     * This animation moves the component to a new position.
     * It uses a linear ease.
     *
     * @param  {string} newPosition coords as x,y string.
     * @return {Promise<void>}      animation's promise.
     */
    updatePosition(newPosition: string): Promise<void> {
      this.position = newPosition;

      const positionNumbers =
        this.position
          .split(',')
          .map((string: string): number => (parseInt(string)));

      return new Promise((onComplete) => {
        TweenLite.to(this.view, movementDuration, {
          x: positionNumbers[0] * (size + margin) + size / HALF + margin,
          y: positionNumbers[1] * (size + margin) + size / HALF + margin,
          ease: Linear.easeNone,
          onComplete,
        });
      });
    },
    /**
     * Reset the actor's position to the initial one.
     * No animation is involved.
     *
     * @returns {void}
     */
    resetPosition() {
      this.position = `${this.initialPosition[0]},${this.initialPosition[1]}`;

      this.view.setParent(initialParent);

      this.view.x = this.initialPosition[0] * (size + margin) + size / HALF + margin;
      this.view.y = this.initialPosition[1] * (size + margin) + size / HALF + margin;
    },
  };
};

export default positioningFunctionalityBuilder;
