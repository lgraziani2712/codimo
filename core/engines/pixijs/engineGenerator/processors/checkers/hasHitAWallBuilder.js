/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
import parseEmoji from 'core/helpers/parseEmoji';
import {
  MOVE_FORWARD,
  MOVE_RIGHT,
  MOVE_BACKWARD,
  MOVE_LEFT,
} from 'core/constants/instructions';
import { type CodimoComponent } from 'core/engines/pixijs/components/componentGenerator';
import { type EngineData } from 'core/engines/pixijs/engineGenerator';

import { type PositioningState } from '../positioningProcessorBuilder';
import { type Checker } from '../processorGenerator';

import engineErrorBuilder from './engineErrorBuilder';
import HasHitAWallError from './HasHitAWallError.gif';

type ActivePathBorders = {|
  bottom?: boolean,
  left?: boolean,
  right?: boolean,
  top?: boolean,
|};
/**
 * The string represents a coord x,y.
 */
export type EngineData$Path = Map<string, ActivePathBorders>;

const directionsToWalls = {
  [MOVE_FORWARD]: 'top',
  [MOVE_RIGHT]: 'right',
  [MOVE_BACKWARD]: 'bottom',
  [MOVE_LEFT]: 'left',
};

const hasHitaWallError = engineErrorBuilder('HasHitAWallError', {
  imageUrl: HasHitAWallError,
  title: parseEmoji('üôÖ POR AH√ç NO üôÖ‚Äç‚ôÇÔ∏è'),
  html: '¬°UY! ¬°TE CHOCASTE LA PARED!<br/>TRAT√Å DE SEGUIR EL CAMINO',
});

/**
 * It validates the next position.
 * If there is a wall, it throws a new exection
 * and stop the execution of the animation.
 *
 * @version 1.0.1
 * @param  {CodimoComponent}  component  The component to check.
 * @param  {EngineData}       engineData Contains the required data for validation.
 * @return {Checker}                     The new instance.
 */
export default function hasHitAWallBuilder(
  component: CodimoComponent,
  engineData: EngineData,
): Checker {
  if (typeof component.hitTheWall !== 'function') {
    throw new Error(
      '`hasHitAWall` checker requires the component to have the `hitTheWall` functionality',
    );
  }
  if (!Array.isArray(engineData.path)) {
    throw new Error(
      // TODO add the URL for the doc of the path shape!
      '`hasHitAWall` checker requires `engineData.path`. ' +
      'See blablabla for the `path` shape.',
    );
  }
  const paths: EngineData$Path = new Map(engineData.path);

  return async ({ instruction, oldPosition }: PositioningState) => {
    // $FlowDoNotDisturb it exists
    const path: ActivePathBorders = paths.get(oldPosition);

    if (!path[directionsToWalls[instruction]]) {
      await component.hitTheWall(directionsToWalls[instruction]);

      throw hasHitaWallError;
    }
  };
}
