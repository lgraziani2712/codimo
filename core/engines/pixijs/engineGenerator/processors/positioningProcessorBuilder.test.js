/**
 * @author Luciano Graziani @lgraziani2712
 * @license {@link http://www.opensource.org/licenses/mit-license.php|MIT License}
 *
 * @flow
 */
/* eslint-disable no-magic-numbers, object-property-newline */
import { Container } from 'pixi.js';

import {
  MOVE_FORWARD,
  MOVE_LEFT,
} from 'core/constants/instructions';
import componentGenerator from 'core/engines/pixijs/components/componentGenerator';
import positioningFunctionalityBuilder
  from 'core/engines/pixijs/components/functionalities/positioningFunctionalityBuilder';
import hitTheWallFunctionalityBuilder
  from 'core/engines/pixijs/components/functionalities/hitTheWallFunctionalityBuilder';

import { positioningProcessorBuilder } from './positioningProcessorBuilder';
import hasHitAWallBuilder from './checkers/hasHitAWallBuilder';

const size = 64;
const margin = 10;
const initialPosition = '1,4';
const positioningMetadata = {
  canvas: {
    height: 569,
    width: 450,
  },
  width: 3,
  height: 5,
  margin: 10,
  size: 64,
  path: [
    ['1,4', { top: true }],
    ['1,3', { top: true, bottom: true }],
    ['1,2', { top: true, bottom: true }],
    ['1,1', { top: true, bottom: true }],
    ['1,0', { top: true, bottom: true }],
  ],
};

describe('engines/pixijs/engineGenerator/processors/positioningProcessorBuilder', () => {
  it('should do nothing if the instruction is unknown', async () => {
    const actorView = new Container();

    actorView.setParent(new Container());

    const actor =
      componentGenerator(actorView, size, margin)
        .addFunctionality(
          'positioning',
          positioningFunctionalityBuilder(initialPosition),
        )
        .build();
    const beforeUpdateStateCheckers = new Map();
    const positioningProcessor = positioningProcessorBuilder(
      actor,
      beforeUpdateStateCheckers,
    );

    await positioningProcessor({
      id: '1',
      key: 'doNothing',
      params: [],
    });

    expect(actor.position).toBe(initialPosition);
  });
  it('should update the position if is a known instruction', async () => {
    const actorView = new Container();

    actorView.setParent(new Container());

    const actor =
      componentGenerator(actorView, size, margin)
        .addFunctionality(
          'positioning',
          positioningFunctionalityBuilder(initialPosition),
        )
        .build();
    const beforeUpdateStateCheckers = new Map();
    const positioningProcessor = positioningProcessorBuilder(
      actor,
      beforeUpdateStateCheckers,
    );

    await positioningProcessor({
      id: '1',
      key: MOVE_FORWARD,
      params: ['1'],
    });

    expect(actor.position).toBe('1,3');
  });
  it('should throw a HasHitAWallError if has hit a wall', async () => {
    const actorView = new Container();

    actorView.setParent(new Container());

    const actor =
      componentGenerator(actorView, size, margin)
        .addFunctionality(
          'positioning',
          positioningFunctionalityBuilder(initialPosition),
        )
        .addFunctionality('hitthewall', hitTheWallFunctionalityBuilder())
        .build();
    const beforeUpdateStateCheckers = new Map([
      ['hashitwall', hasHitAWallBuilder(actor, positioningMetadata)],
    ]);
    const positioningProcessor = positioningProcessorBuilder(
      actor,
      beforeUpdateStateCheckers,
    );

    try {
      await positioningProcessor({
        id: '1',
        key: MOVE_LEFT,
        params: ['1'],
      });
    } catch (error) {
      expect(error.name).toBe('HasHitAWallError');
      expect(error.html).toMatchSnapshot();

      return;
    }

    expect('true').toBe('not evaluated');
  });
});
