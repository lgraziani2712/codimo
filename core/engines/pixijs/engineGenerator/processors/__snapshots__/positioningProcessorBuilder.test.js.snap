// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engines/pixijs/engineGenerator/processors/positioningProcessorBuilder should throw a HasHitAWallError if has hit a wall 1`] = `"La dirección que intentás tomar es incorrecta"`;
